### PR TITLE
Etc ficus create enroll api

### PR DESCRIPTION
**Pull request by AndreyLykhoman**
**Description:**  Endpoints fo create user and enroll user. Also add endpoint for enable and disable users 

For this task I used solving relates use used the workpiece. I configurated setting and wrote needed endpoints.

**Youtrack:** https://youtrack.raccoongang.com/issue/ETC-11

**Configuration instructions:** 
Code need next additional settings:

add to lms.auth.json 'EDX_API_KEY'. It's key for check valid token.
The following token should be available in the python code: 'HTTP_X_EDX_API_KEY,
X-Edx-Api-Key HTTP header is present in the request

**Testing instructions:**
endpoints:

'api/user/create' - body { Prename, surname, email, username} - Response {user_id}
curl -X POST
http://localhost:8000/api/user/create
-H 'Cache-Control: no-cache'
-H 'Content-Type: application/json'
-H 'X-Edx-Api-Key: PUT_YOUR_API_KEY_HERE'
-H 'content-type: multipart/form-data; boundary=----WebKitFormBoundary7MA4YWxkTrZu0gW'
-F email=andrey.lykhoman12344566@raccoongang.com
-F username=andreylykhoman006
-F prename=Andrey
-F surname=Lykhoman

-'api/user/set/activate/status' - body {user_id, is_active} - Response {user_id, is_active}
curl -X POST
http://localhost:8000/api/user/set/activate/status
-H 'Cache-Control: no-cache'
-H 'X-Edx-Api-Key: PUT_YOUR_API_KEY_HERE'
-H 'content-type: multipart/form-data; boundary=----WebKitFormBoundary7MA4YWxkTrZu0gW'
-F user_id=21
-F is_active=true

'/api/user/enrollment' - body {User_id, course_id, mode, is_active} - Response {enroll_id}
curl -X POST
http://localhost:8000/api/user/enrollment
-H 'Cache-Control: no-cache'
-H 'Content-Type: application/x-www-form-urlencoded'
-H 'X-Edx-Api-Key: PUT_YOUR_API_KEY_HERE'
-H 'content-type: multipart/form-data; boundary=----WebKitFormBoundary7MA4YWxkTrZu0gW'
-F 'course_id=course-v1:edX+DemoX+Demo_Course'
-F user_id=21
-F is_active=false
-F mode=honor

### CODE REVIEW

File: cms/envs/appsembler.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+from path import path
+
+SERVICE_VARIANT = os.environ.get('SERVICE_VARIANT', None)
+CONFIG_ROOT = path('/edx/app/edxapp/')  # don't hardcode this in the future
+CONFIG_PREFIX = SERVICE_VARIANT + "." if SERVICE_VARIANT else ""
+with open(CONFIG_ROOT / CONFIG_PREFIX + 'env.json') as env_file:
+    ENV_TOKENS = json.load(env_file)
+
+APPSEMBLER_FEATURES = ENV_TOKENS.get('APPSEMBLER_FEATURES', {})
```
Comment (idegtiarov) [inline cms/envs/appsembler.py]: Why do we have appsembler.py and all this variable vendor staff in our code?

File: common/djangoapps/etcapi/context_processors.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
+import hmac
+import hashlib
+import os
+
+from django.conf import settings
+
+from student import roles
+
+
+def intercom(request):
```
Comment (idegtiarov) [inline common/djangoapps/etcapi/context_processors.py]: Where is the function's description?

File: common/djangoapps/etcapi/context_processors.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    data = {'show_intercom_widget': False}
+
+    intercom_app_id = os.environ.get("INTERCOM_APP_ID", "")
+    if not intercom_app_id:
+        return data
+
+    if settings.PROJECT_ROOT.endswith('edx-platform/cms') or \
```
Comment (idegtiarov) [inline common/djangoapps/etcapi/context_processors.py]: Back-slash as a new line separator, really?! Please take a look at code styling http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/style_guides/python-guidelines.html#id4

File: common/djangoapps/etcapi/models.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+from django.contrib.auth.models import User
+
+from intercom import intercom
+import requests
+
+from xmodule.modulestore.django import modulestore
+from student.models import CourseEnrollment
+from course_action_state.models import CourseRerunState
+
+# try:
```
Comment (idegtiarov) [inline common/djangoapps/etcapi/models.py]: Please remove unused strings

File: lms/djangoapps/etc_api/serializers.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        if type(data) is list or type(data) is tuple:
+            data = data[0]
+        data_list = data.split(',')
+        without_null_string_data_list = []
+        for course in data_list:
+            if len(course) > 0:
+                without_null_string_data_list.append(course)
+        return without_null_string_data_list
+
+class BulkEnrollmentSerializer(serializers.Serializer):
```
Comment (idegtiarov) [inline lms/djangoapps/etc_api/serializers.py]: Why do we still need bulk enrollment? Description?

File: lms/djangoapps/etc_api/serializers.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
+from opaque_keys import InvalidKeyError
+from opaque_keys.edx.keys import CourseKey
+from rest_framework import serializers
+
+
+class StringListField(serializers.ListField):
```
Comment (idegtiarov) [inline lms/djangoapps/etc_api/serializers.py]: Where is a description of the class?

File: lms/djangoapps/etc_api/utils.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+        return False
+
+class ApiKeyHeaderPermissionInToken(ApiKeyHeaderPermission, permissions.IsAuthenticated):
+    """
+    Custom class to check permission by token
+    """
+
+    def has_permission(self, request, view):
+        """
+               Check for permissions by matching the configured API key and header
```
Comment (idegtiarov) [inline lms/djangoapps/etc_api/utils.py]: Indentation is wrong, please improve

File: lms/djangoapps/etc_api/utils.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+                      'email_from_address', settings.DEFAULT_FROM_EMAIL),
+                  request=request,
+                  subject_template_name='etc_api/set_password_subject.txt',
+                  email_template_name='etc_api/set_password_email.html'
+        )
+        return True
+    else:
+        return False
+
+class ApiKeyHeaderPermissionInToken(ApiKeyHeaderPermission, permissions.IsAuthenticated):
```
Comment (idegtiarov) [inline lms/djangoapps/etc_api/utils.py]: Why don't you use ApiKeyHeaderPermission and create the new one?

File: lms/djangoapps/etc_api/utils.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    """
+    Custom class to check permission by token
+    """
+
+    def has_permission(self, request, view):
+        """
+               Check for permissions by matching the configured API key and header
+               If settings.DEBUG is True and settings.EDX_APP_SEMBLER_API_KEY is not set or None,
+               then allow the request. Otherwise, allow the request if and only if
+               settings.EDX_APP_SEMBLER_API_KEY is set and the X-Edx-App-Semb-Api-Key HTTP header is
```
Comment (idegtiarov) [inline lms/djangoapps/etc_api/utils.py]: APP_SEMMLER?

File: lms/djangoapps/etc_api/views.py [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+
+from django.core.exceptions import NON_FIELD_ERRORS, ValidationError
+from django.contrib.auth.models import User
+from rest_framework.views import APIView
+from rest_framework import status
+from rest_framework.response import Response
+from util.disable_rate_limit import can_disable_rate_limit
+from openedx.core.djangoapps.user_api.accounts.api import check_account_exists
+from student.views import create_account_with_params
+from student.models import CourseEnrollment, EnrollmentClosedError, \
```
Comment (idegtiarov) [inline lms/djangoapps/etc_api/views.py]: Please take a look at edx code style guide!

File: lms/djangoapps/etc_api/views.py [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+from instructor.views.api import save_registration_code, \
+    students_update_enrollment, require_level
+from .serializers import BulkEnrollmentSerializer
+from .utils import  send_activation_email, ApiKeyHeaderPermissionInToken
+from django.contrib.auth import authenticate, login
+from rest_framework.authentication import SessionAuthentication, BasicAuthentication
+from django.core.validators  import validate_slug
+from opaque_keys.edx.locations import SlashSeparatedCourseKey
+
+class CsrfExemptSessionAuthentication(SessionAuthentication):
```
Comment (idegtiarov) [inline lms/djangoapps/etc_api/views.py]: Why do you need this class? Could you please explain?

File: lms/djangoapps/etc_api/views.py [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+
+
+
+class CreateUserAccountWithoutPasswordView(APIView):
+    authentication_classes = (CsrfExemptSessionAuthentication, BasicAuthentication)
+    permission_classes = ApiKeyHeaderPermissionInToken,
+
+
+    def post(self, request):
+        """
```
Comment (idegtiarov) [inline lms/djangoapps/etc_api/views.py]: It is the great doc string!

File: lms/djangoapps/etc_api/views.py [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+
+        """
+        #{Prename, surname, email, username}
```
Comment (idegtiarov) [inline lms/djangoapps/etc_api/views.py]: Strange comment, need to be improved or removed!

File: lms/djangoapps/etc_api/views.py [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+        data = request.data
+        # set the honor_code and honor_code like checked,
```
Comment (idegtiarov) [inline lms/djangoapps/etc_api/views.py]: `honor_code and honor_code`???

File: lms/djangoapps/etc_api/views.py [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+            errors = {"user_message": "User already exists"}
+            return Response(errors, status=409)
+
+        try:
+            password = ''.join(random.choice(
+                string.ascii_uppercase + string.ascii_lowercase + string.digits)
+                               for _ in range(32))
+
+            data['password'] = password
+            data['send_activation_email'] = False
```
Comment (idegtiarov) [inline lms/djangoapps/etc_api/views.py]: why is this parameter added to the data?

File: lms/djangoapps/etc_api/views.py [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+
+        return response
+
+class SetActivateUserStatus(APIView):
+    authentication_classes = (CsrfExemptSessionAuthentication, BasicAuthentication)
+    permission_classes = ApiKeyHeaderPermissionInToken,
+
+
+    def post(self, request):
+        """
```
Comment (idegtiarov) [inline lms/djangoapps/etc_api/views.py]: ditto

File: lms/djangoapps/etc_api/views.py [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+
+        """
+        #{user_id, activate status}
```
Comment (idegtiarov) [inline lms/djangoapps/etc_api/views.py]: ditto

File: lms/djangoapps/etc_api/views.py [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+        data = request.data
+        # set the honor_code and honor_code like checked,
```
Comment (idegtiarov) [inline lms/djangoapps/etc_api/views.py]: What does it mean?

File: lms/djangoapps/etc_api/views.py [hunk 19]
```diff
@@ -0,0 +1,178 @@
...
+            errors = {"user_message": "Wrong id User does not exist"}
+            return Response(errors, status=400)
+
+        response = Response({'user_id': user_id, 'is_active': is_active}, status=200)
+
+        return response
+
+
+@can_disable_rate_limit
+class BulkEnrollView(APIView, ApiKeyPermissionMixIn):
```
Comment (idegtiarov) [inline lms/djangoapps/etc_api/views.py]: why do we use bulk enrollment?

File: lms/djangoapps/etc_api/views.py [hunk 20]
```diff
@@ -0,0 +1,178 @@
...
+    authentication_classes = (CsrfExemptSessionAuthentication, BasicAuthentication)
+    permission_classes = ApiKeyHeaderPermissionInToken,
+
+    def post(self, request):
+
+        data = request.data
+        for key, value in data.iteritems():
+            data[key] = str(value)
+        #user_id = int(data['user_id'])
```
Comment (idegtiarov) [inline lms/djangoapps/etc_api/views.py]: ?

File: openedx/core/djangoapps/user_api/urls.py [hunk 21]
```diff
@@ -48,4 +51,17 @@
         PreferencesDetailView.as_view(),
         name='preferences_detail_api'
     ),
+    url(
```
Comment (idegtiarov) [inline openedx/core/djangoapps/user_api/urls.py]: Could these urls match one line?

File: lms/djangoapps/etc_api/views.py [hunk 22]
```diff
@@ -0,0 +1,178 @@
...
+    def post(self, request):
+        """
+
+        """
+        #{user_id, activate status}
+        data = request.data
+        # set the honor_code and honor_code like checked,
+        # so we can use the already defined methods for creating an user
+        try:
+            user_id = int(data['user_id'])
```
Comment (idegtiarov) [inline lms/djangoapps/etc_api/views.py]: Do we need this operation?

File: lms/djangoapps/etc_api/views.py [hunk 23]
```diff
@@ -0,0 +1,178 @@
...
+        except  User.DoesNotExist:
+            errors = {"user_message": "Wrong id User does not exist"}
+            return Response(errors, status=400)
+
+        response = Response({'user_id': user_id, 'is_active': is_active}, status=200)
+
+        return response
+
+
+@can_disable_rate_limit
```
Comment (idegtiarov) [inline lms/djangoapps/etc_api/views.py]: Why we use this decorator?

Comment (idegtiarov) [reply lms/djangoapps/etc_api/views.py]: In first PR I saw tests where are they now?

**Comment by idegtiarov** [comment]
In first PR I saw tests where are they now?

File: lms/djangoapps/etc_api/utils.py (hunk 8 above)
Comment (AndreyLykhoman) [reply lms/djangoapps/etc_api/utils.py]: It's need for checking permission and checking enable or disable api for current site.

File: lms/djangoapps/etc_api/views.py (hunk 11 above)
Comment (AndreyLykhoman) [reply lms/djangoapps/etc_api/views.py]: This class allows you to turn off checking the security token for the mappings where it is used.

File: openedx/core/djangoapps/user_api/urls.py [hunk 24]
```diff
@@ -48,4 +51,8 @@
         PreferencesDetailView.as_view(),
         name='preferences_detail_api'
     ),
+    url(r'create$', CreateUserAccountWithoutPasswordView.as_view()),
+    url(r'enrollment$',EnrollView.as_view()),
```
Comment (flying-pi) [inline openedx/core/djangoapps/user_api/urls.py]: Add some space

File: openedx/core/djangoapps/user_api/urls.py [hunk 25]
```diff
@@ -48,4 +51,8 @@
...
+    url(r'set/activate/status$',SetActivateUserStatus.as_view()),
```
Comment (flying-pi) [inline openedx/core/djangoapps/user_api/urls.py]: Add some space

File: openedx/core/djangoapps/user_api/urls.py [hunk 26]
```diff
@@ -9,7 +9,10 @@
 from .accounts.views import AccountViewSet
 from .preferences.views import PreferencesView, PreferencesDetailView
 from .verification_api.views import PhotoVerificationStatusView
-
+from lms.djangoapps.etc_api.views import (
```
Comment (flying-pi) [inline openedx/core/djangoapps/user_api/urls.py]: Can we join it in single line?

File: lms/djangoapps/etc_api/views.py [hunk 27]
```diff
@@ -0,0 +1,136 @@
...
+        except  User.DoesNotExist:
+            errors = {"user_message": "Wrong id User does not exist"}
+            return Response(errors, status=400)
+        course_id = SlashSeparatedCourseKey.from_deprecated_string(data['course_id'])
+        enrollment_obj, __ = CourseEnrollment.objects.update_or_create(
+            user=user,
+            course_id=course_id,
+            defaults={
+                'mode': data['mode'],
+                'is_active': True if 'true' == "{}".format(data['is_active']).lower() else False
```
Comment (flying-pi) [inline lms/djangoapps/etc_api/views.py]: move to function with condition like 
```
str(value).lower() in ['true',]
```

File: lms/djangoapps/etc_api/views.py [hunk 28]
```diff
@@ -0,0 +1,136 @@
...
+            return Response(errors, status=409)
+
+        try:
+            data['password'] = uuid4().hex
+            user = create_account_with_params(request, data)
+            user.is_active = True
+            user.first_name = request.data.get('prename')
+            user.last_name = request.data.get('surname')
+            user.save()
+            user_id = user.id
```
Comment (flying-pi) [inline lms/djangoapps/etc_api/views.py]: Why you create this var

File: lms/djangoapps/etc_api/views.py [hunk 29]
```diff
@@ -0,0 +1,136 @@
...
+            self.send_activation_email(request)
+        except ValidationError as err:
+            # Should only get non-field errors from this function
+            assert NON_FIELD_ERRORS not in err.message_dict
+            # Only return first error for each field
+            errors = {"user_message": "Wrong parameters on user creation"}
```
Comment (flying-pi) [inline lms/djangoapps/etc_api/views.py]: Why you need this var?

File: lms/djangoapps/etc_api/views.py [hunk 30]
```diff
@@ -0,0 +1,136 @@
...
+            return Response(errors, status=400)
+        except ValueError as err:
+            errors = {"user_message": "Wrong email format"}
```
Comment (flying-pi) [inline lms/djangoapps/etc_api/views.py]: Why you need this var?

File: lms/djangoapps/etc_api/views.py [hunk 31]
```diff
@@ -0,0 +1,136 @@
...
+        form = PasswordResetFormNoActive(request.data)
+        if form.is_valid():
+            form.save(use_https=request.is_secure(),
+                      from_email=configuration_helpers.get_value(
+                          'email_from_address', settings.DEFAULT_FROM_EMAIL),
+                      request=request,
+                      subject_template_name='etc_api/set_password_subject.txt',
+                      email_template_name='etc_api/set_password_email.html'
+                      )
+            return True
```
Comment (flying-pi) [inline lms/djangoapps/etc_api/views.py]: The result of this function doesn't use. Do we need it?

**Comment by AndreyLykhoman** [comment]
@idegtiarov  
I deleted the tests because there was a test of the code that was deleted.

**Pull request merged**